### ISSUE TITLE
Feature Request: Tree shake-able class transform

### ISSUE THREAD

**Issue by pshrmn** [babel/babel#5983]
Choose one: is this a bug report or feature request? **feature request**

### Input Code

```js
class MyComponent extends React.Component {
  // ...
}
```

### Babel Configuration (.babelrc, package.json, cli command)

```js
{
  "plugins": [ [ "transform-es2015-classes", { inline: true, loose: true } ] ]
}
```

### Expected Behavior

Ideally, instead of inserting an `inherits` function into the scope, the code from the `inherits` function would be inserted into the compiled class function

```js
var MyComponent = function(_reactComponent) {
  MyComponent.prototype = Object.create(_reactComponent.prototype);
  MyComponent.prototype.constructor = MyComponent;
  MyComponent.__proto__ = _reactComponent;

  function MyComponent { ... };

  return MyComponent;
}(React.Component);
```
My basic testing (using the Rollup REPL) seems to imply that that would successfully be tree shaken. However, that is only true for the loose inherits, the full inherits still would not be tree shaken.

### Current Behavior

An `inherits` function is inserted into the scope and called within the compiled class.

```js
var inherits = function(subClass, superClass) { ... };

var MyComponent = function(_reactComponent) {
  inherits(MyComponent, _reactComponent);

  function MyComponent { ... };

  return MyComponent;
}(React.Component);
```

### Possible Solution

We could pass an `inline: true` option to the `transform-es2015-classes` plugin and it would inline the inherits code instead of inserting the helper function into the scope and calling that.

I'm not familiar enough with Babel to say whether adding an option to the plugin is a viable solution. Instead, this might have to be done with a separate plugin.

### Context

I was experimenting with moving a project (React Router) to a single file build. Currently, our `main` and `module` builds are just the compiled files from our source. This allows users to cherry pick imports so that their bundles only include the files that they want. If we were to switch to a single file, users obviously could no longer cherry pick, so we would want to make our exports tree shake-able to have feature parity.

This isn't a React specific problem, but since React now encourages using `class <C> extends React.Component`, any React component package that does so will have this issue.

**Comment by hzoo** [comment]
The tree shaking behavior I think is covered by this request: https://github.com/babel/babel/issues/5632 (adding an annotation which I believe to be `/* @class */` based on what typescript did to cover that. otherwise we could make the loose mode always inline as well if we don't want to introduce a new option

**Comment by pshrmn** [comment]
It is interesting that Webpack (or more specifically UglifyJS) cannot eliminate regular compiled classes. I have been doing my testing with Rollup and [regular classes are tree shaken](https://rollupjs.org/repl?version=0.45.2&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QiUyMEZvbyUyMCU3RCUyMGZyb20lMjAnLiUyRnRoaW5ncy5qcyclM0IlNUNuY29uc3QlMjBmJTIwJTNEJTIwbmV3JTIwRm9vKDUpJTNCJTVDbmNvbnNvbGUubG9nKCUyMGYudG9TdHJpbmcoKSUyMCklM0IlMjIlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIydGhpbmdzLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMiU1QyUyMnVzZSUyMHN0cmljdCU1QyUyMiUzQiU1Q24lNUNuZnVuY3Rpb24lMjBfY2xhc3NDYWxsQ2hlY2soaW5zdGFuY2UlMkMlMjBDb25zdHJ1Y3RvciklMjAlN0IlMjBpZiUyMCghKGluc3RhbmNlJTIwaW5zdGFuY2VvZiUyMENvbnN0cnVjdG9yKSklMjAlN0IlMjB0aHJvdyUyMG5ldyUyMFR5cGVFcnJvciglNUMlMjJDYW5ub3QlMjBjYWxsJTIwYSUyMGNsYXNzJTIwYXMlMjBhJTIwZnVuY3Rpb24lNUMlMjIpJTNCJTIwJTdEJTIwJTdEJTVDbiU1Q252YXIlMjBGb28lMjAlM0QlMjBmdW5jdGlvbiUyMCgpJTIwJTdCJTVDbiUyMCUyMGZ1bmN0aW9uJTIwRm9vKG51bSklMjAlN0IlNUNuJTIwJTIwJTIwJTIwX2NsYXNzQ2FsbENoZWNrKHRoaXMlMkMlMjBGb28pJTNCJTVDbiU1Q24lMjAlMjAlMjAlMjB0aGlzLm51bSUyMCUzRCUyMG51bSUzQiU1Q24lMjAlMjAlN0QlNUNuJTVDbiUyMCUyMEZvby5wcm90b3R5cGUudG9TdHJpbmclMjAlM0QlMjBmdW5jdGlvbiUyMHRvU3RyaW5nKCklMjAlN0IlNUNuJTIwJTIwJTIwJTIwcmV0dXJuJTIwJTVDJTIyJTNDRm9vJTNFJTIwJTVDJTIyJTIwJTJCJTIwdGhpcy5udW0lM0IlNUNuJTIwJTIwJTdEJTNCJTVDbiU1Q24lMjAlMjByZXR1cm4lMjBGb28lM0IlNUNuJTdEKCklM0IlNUNuJTVDbnZhciUyMEJhciUyMCUzRCUyMGZ1bmN0aW9uJTIwKCklMjAlN0IlNUNuJTIwJTIwZnVuY3Rpb24lMjBCYXIobnVtKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjBfY2xhc3NDYWxsQ2hlY2sodGhpcyUyQyUyMEJhciklM0IlNUNuJTVDbiUyMCUyMCUyMCUyMHRoaXMubnVtJTIwJTNEJTIwbnVtJTNCJTVDbiUyMCUyMCU3RCU1Q24lNUNuJTIwJTIwQmFyLnByb3RvdHlwZS50b1N0cmluZyUyMCUzRCUyMGZ1bmN0aW9uJTIwdG9TdHJpbmcoKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjByZXR1cm4lMjAlNUMlMjIlM0NCYXIlM0UlMjAlNUMlMjIlMjAlMkIlMjB0aGlzLm51bSUzQiU1Q24lMjAlMjAlN0QlM0IlNUNuJTVDbiUyMCUyMHJldHVybiUyMEJhciUzQiU1Q24lN0QoKSUzQiU1Q24lNUNuZXhwb3J0JTIwJTdCJTIwRm9vJTJDJTIwQmFyJTIwJTdEJTNCJTVDbiUyMiU3RCU1RCUyQyUyMm9wdGlvbnMlMjIlM0ElN0IlMjJmb3JtYXQlMjIlM0ElMjJjanMlMjIlMkMlMjJtb2R1bGVOYW1lJTIyJTNBJTIybXlCdW5kbGUlMjIlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTJDJTIybW9kdWxlSWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJleGFtcGxlJTIyJTNBbnVsbCU3RA==), only failing to tree shake [classes that use `extends`](https://rollupjs.org/repl?version=0.45.2&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QiUyMFBhcmVudCUyMCU3RCUyMGZyb20lMjAnLiUyRnRoaW5ncy5qcyclM0IlNUNuY29uc3QlMjBwJTIwJTNEJTIwbmV3JTIwUGFyZW50KDUpJTNCJTVDbmNvbnNvbGUubG9nKCUyMHAudG9TdHJpbmcoKSUyMCklM0IlMjIlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIydGhpbmdzLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMiU1QyUyMnVzZSUyMHN0cmljdCU1QyUyMiUzQiU1Q24lNUNuZnVuY3Rpb24lMjBfcG9zc2libGVDb25zdHJ1Y3RvclJldHVybihzZWxmJTJDJTIwY2FsbCklMjAlN0IlMjBpZiUyMCghc2VsZiklMjAlN0IlMjB0aHJvdyUyMG5ldyUyMFJlZmVyZW5jZUVycm9yKCU1QyUyMnRoaXMlMjBoYXNuJ3QlMjBiZWVuJTIwaW5pdGlhbGlzZWQlMjAtJTIwc3VwZXIoKSUyMGhhc24ndCUyMGJlZW4lMjBjYWxsZWQlNUMlMjIpJTNCJTIwJTdEJTIwcmV0dXJuJTIwY2FsbCUyMCUyNiUyNiUyMCh0eXBlb2YlMjBjYWxsJTIwJTNEJTNEJTNEJTIwJTVDJTIyb2JqZWN0JTVDJTIyJTIwJTdDJTdDJTIwdHlwZW9mJTIwY2FsbCUyMCUzRCUzRCUzRCUyMCU1QyUyMmZ1bmN0aW9uJTVDJTIyKSUyMCUzRiUyMGNhbGwlMjAlM0ElMjBzZWxmJTNCJTIwJTdEJTVDbiU1Q25mdW5jdGlvbiUyMF9pbmhlcml0cyhzdWJDbGFzcyUyQyUyMHN1cGVyQ2xhc3MpJTIwJTdCJTIwaWYlMjAodHlwZW9mJTIwc3VwZXJDbGFzcyUyMCElM0QlM0QlMjAlNUMlMjJmdW5jdGlvbiU1QyUyMiUyMCUyNiUyNiUyMHN1cGVyQ2xhc3MlMjAhJTNEJTNEJTIwbnVsbCklMjAlN0IlMjB0aHJvdyUyMG5ldyUyMFR5cGVFcnJvciglNUMlMjJTdXBlciUyMGV4cHJlc3Npb24lMjBtdXN0JTIwZWl0aGVyJTIwYmUlMjBudWxsJTIwb3IlMjBhJTIwZnVuY3Rpb24lMkMlMjBub3QlMjAlNUMlMjIlMjAlMkIlMjB0eXBlb2YlMjBzdXBlckNsYXNzKSUzQiUyMCU3RCUyMHN1YkNsYXNzLnByb3RvdHlwZSUyMCUzRCUyME9iamVjdC5jcmVhdGUoc3VwZXJDbGFzcyUyMCUyNiUyNiUyMHN1cGVyQ2xhc3MucHJvdG90eXBlJTJDJTIwJTdCJTIwY29uc3RydWN0b3IlM0ElMjAlN0IlMjB2YWx1ZSUzQSUyMHN1YkNsYXNzJTJDJTIwZW51bWVyYWJsZSUzQSUyMGZhbHNlJTJDJTIwd3JpdGFibGUlM0ElMjB0cnVlJTJDJTIwY29uZmlndXJhYmxlJTNBJTIwdHJ1ZSUyMCU3RCUyMCU3RCklM0IlMjBpZiUyMChzdXBlckNsYXNzKSUyME9iamVjdC5zZXRQcm90b3R5cGVPZiUyMCUzRiUyME9iamVjdC5zZXRQcm90b3R5cGVPZihzdWJDbGFzcyUyQyUyMHN1cGVyQ2xhc3MpJTIwJTNBJTIwc3ViQ2xhc3MuX19wcm90b19fJTIwJTNEJTIwc3VwZXJDbGFzcyUzQiUyMCU3RCU1Q24lNUNuZnVuY3Rpb24lMjBfY2xhc3NDYWxsQ2hlY2soaW5zdGFuY2UlMkMlMjBDb25zdHJ1Y3RvciklMjAlN0IlMjBpZiUyMCghKGluc3RhbmNlJTIwaW5zdGFuY2VvZiUyMENvbnN0cnVjdG9yKSklMjAlN0IlMjB0aHJvdyUyMG5ldyUyMFR5cGVFcnJvciglNUMlMjJDYW5ub3QlMjBjYWxsJTIwYSUyMGNsYXNzJTIwYXMlMjBhJTIwZnVuY3Rpb24lNUMlMjIpJTNCJTIwJTdEJTIwJTdEJTVDbiU1Q252YXIlMjBQYXJlbnQlMjAlM0QlMjBmdW5jdGlvbiUyMCgpJTIwJTdCJTVDbiUyMCUyMGZ1bmN0aW9uJTIwUGFyZW50KG51bSklMjAlN0IlNUNuJTIwJTIwJTIwJTIwX2NsYXNzQ2FsbENoZWNrKHRoaXMlMkMlMjBQYXJlbnQpJTNCJTVDbiU1Q24lMjAlMjAlMjAlMjB0aGlzLm51bSUyMCUzRCUyMG51bSUzQiU1Q24lMjAlMjAlN0QlNUNuJTVDbiUyMCUyMFBhcmVudC5wcm90b3R5cGUudG9TdHJpbmclMjAlM0QlMjBmdW5jdGlvbiUyMHRvU3RyaW5nKCklMjAlN0IlNUNuJTIwJTIwJTIwJTIwcmV0dXJuJTIwJTVDJTIyJTNDUGFyZW50JTNFJTIwJTVDJTIyJTIwJTJCJTIwdGhpcy5udW0lM0IlNUNuJTIwJTIwJTdEJTNCJTVDbiU1Q24lMjAlMjByZXR1cm4lMjBQYXJlbnQlM0IlNUNuJTdEKCklM0IlNUNuJTVDbnZhciUyMENoaWxkJTIwJTNEJTIwZnVuY3Rpb24lMjAoX1BhcmVudCklMjAlN0IlNUNuJTIwJTIwX2luaGVyaXRzKENoaWxkJTJDJTIwX1BhcmVudCklM0IlNUNuJTVDbiUyMCUyMGZ1bmN0aW9uJTIwQ2hpbGQobnVtKSUyMCU3QiU1Q24lMjAlMjAlMjAlMjBfY2xhc3NDYWxsQ2hlY2sodGhpcyUyQyUyMENoaWxkKSUzQiU1Q24lNUNuJTIwJTIwJTIwJTIwcmV0dXJuJTIwX3Bvc3NpYmxlQ29uc3RydWN0b3JSZXR1cm4odGhpcyUyQyUyMF9QYXJlbnQuY2FsbCh0aGlzJTJDJTIwbnVtKSklM0IlNUNuJTIwJTIwJTdEJTVDbiU1Q24lMjAlMjBDaGlsZC5wcm90b3R5cGUudG9TdHJpbmclMjAlM0QlMjBmdW5jdGlvbiUyMHRvU3RyaW5nKCklMjAlN0IlNUNuJTIwJTIwJTIwJTIwcmV0dXJuJTIwJTVDJTIyJTNDUGFyZW50JTNFJTIwJTVDJTIyJTIwJTJCJTIwdGhpcy5udW0lM0IlNUNuJTIwJTIwJTdEJTNCJTVDbiU1Q24lMjAlMjByZXR1cm4lMjBDaGlsZCUzQiU1Q24lN0QoUGFyZW50KSUzQiU1Q24lNUNuZXhwb3J0JTIwJTdCJTIwUGFyZW50JTJDJTIwQ2hpbGQlMjAlN0QlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyY2pzJTIyJTJDJTIybW9kdWxlTmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCUyQyUyMm1vZHVsZUlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=). Rollup does fail to tree shake when not in loose mode, but inlining also relies on loose mode, so that wasn't a big deal to me. I'd be curious to see why UglifyJS sees the compiled classes as having side effects while Rollup does not.

If everything could eliminate unused classes, then the inlining would sound better to me. I'm assuming that Uglify isn't likely changing how they identify side effects, so since that will still fail with inlined classes, it seems like the annotation is possibly the way that this would have to go. I know that there was some discussion on adding [pure support to Rollup](https://github.com/rollup/rollup/issues/1293), although I don't think that that has been actively developed.

I think that this could probably be closed in favor of #5632.

**Comment by hzoo** [comment]
Duplicate of #5632.

Ok closing in favor of https://github.com/babel/babel/issues/5632, feel free to add your comments there too.
